### ISSUE TITLE
info pointed by catalog URL makes rancher crash at ~2 minutes interval

### ISSUE THREAD

**Issue by fgeorgatos** [rancher/rancher#19279]
**What kind of request is this (question/bug/enhancement/feature request):**

bug: unexpected container crash (i.e. a heavy slowdown)

**Steps to reproduce (least amount of steps as possible):**

add catalog https://swissdatasciencecenter.github.io/helm-charts/

**Result:**

DoS, with this log visible:

```
2019/03/30 00:16:03 [ERROR] CatalogController renku-helm-charts [catalog] failed with : failed to update templates. Multiple error occurred: [CatalogTemplateVersion.management.cattle.io "renku-helm-charts-renku-notebooks-0.2.1-gputest-" is invalid: metadata.name: Invalid value: "renku-helm-charts-renku-notebooks-0.2.1-gputest-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')] | stdout
-- | --
```

**Other details that may be helpful:**

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.1
- Installation option (single install/HA): single install

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): VM
- Kubernetes version (use `kubectl version`): 1.11.9 & 1.13.5 - verified in both

```
2019/03/30 00:16:03 [ERROR] CatalogController renku-helm-charts [catalog] failed with : failed to update templates. Multiple error occurred: [CatalogTemplateVersion.management.cattle.io "renku-helm-charts-renku-notebooks-0.2.1-gputest-" is invalid: metadata.name: Invalid value: "renku-helm-charts-renku-notebooks-0.2.1-gputest-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')] | stdout
-- | --
```

- Docker version (use `docker version`):

```
18.09.04, docker-ce@centos
```

**Comment by mrajashree** [comment]
@fgeorgatos With container crash you are referring to this CatalogController error right? Your chart name from this catalog seems to be ending with a `-` and not an alphanumeric character as seen in the error log `renku-helm-charts-renku-notebooks-0.2.1-gputest-`, is that right? That should be changed

**Comment by deniseschannon** [comment]
@tfiduccia can you test with the helm chart and let me know if all of rancher crashes or what is it that "crashes"?

**Comment by tfiduccia** [comment]
@deniseschannon Not seeing a "crash".  But adding that catalog seem to be causing some sort of download recursion.  It just keeps trying to sync charts that it already sync'ed.  At some point I'm sure I would run out of memory or something cause it just doesn't stop.

I did add a chart to my helm catalog with a - at the end of it.  It didn't crash, it just didn't show up when I went to apps.

**Comment by deniseschannon** [comment]
@tfiduccia it sounds like something different from helm chart repos versus git repos. 

we should investigate to prevent the recursion. @mrajashree

**Comment by fgeorgatos** [comment]
@tfiduccia : thanks for looking into this. Any news? 
just to confirm, the word "crash" was probably misnomer for saying "unresponsive". I cannot recall what I had to do to rescue it, but it was non-trivial enough to force me to report it here.

**Comment by fgeorgatos** [comment]
@alena1108 , @tfiduccia : indeed, `renku-notebooks-0.2.1-gputest-` is a bug and should not have been produced. Could you make it more defensive from rancher's end (or yield a warning), so that the potential endless loop and memory leak isn't easily feasible?

**Comment by fgeorgatos** [comment]
@tfiduccia could you have another look? 
came across it today again. It's probably not a blocker, but it's an indication of poor defences on behalf of rancher in relation to 3rd party charts...

**Comment by deniseschannon** [comment]
@brendarearden Is it possible that your new fixes around how to handle invalid URLs/charts could fix this?

**Comment by brendarearden** [comment]
@deniseschannon yes, it sounds like it would!

**Comment by deniseschannon** [comment]
@brendarearden can you test out their example repo in this issue and if it's fixed, can you assign to test?

**Comment by jiaqiluo** [comment]
The bug is reproduced in v2.3.5

After adding the abovementioned catalog, Rancher UI shows there is something wrong with the catalog 

<img width="1293" alt="Screen Shot 2020-04-06 at 4 23 14 PM" src="https://user-images.githubusercontent.com/6218999/78614366-395ff580-7823-11ea-8ef0-f033f3216c35.png">


and Rancher logs shows the following repeatedly 
```
2020/04/06 23:22:44 [INFO] Updating project catalog test
2020/04/06 23:22:44 [INFO] Catalog sync done. 0 templates created, 0 templates updated, 0 templates deleted, 2 templates failed
2020/04/06 23:22:44 [ERROR] ProjectCatalogController p-kcc4b/test [projectCatalog] failed with : failed to update templates. Multiple error(s) occurred: [CatalogTemplateVersion.management.cattle.io "test-renku-notebooks-0.2.1-gputest-" is invalid: metadata.name: Invalid value: "test-renku-notebooks-0.2.1-gputest-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*') CatalogTemplateVersion.management.cattle.io "test-renku-ui-0.6.4+038.geebda86" is invalid: metadata.name: Invalid value: "test-renku-ui-0.6.4+038.geebda86": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]
2020-04-06 23:22:47.991332 W | etcdserver: read-only range request "key:\"/registry/management.cattle.io/nodes\" range_end:\"/registry/management.cattle.io/nodet\" count_only:true " with result "range_response_count:0 size:7" took too long (255.40805ms) to execute
2020-04-06 23:22:47.991383 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-scheduler\" " with result "range_response_count:1 size:537" took too long (314.788755ms) to execute
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in Rancher:`v2.4-head 481cb64b3`  and `master-head ea9700b41`

Steps:
- add the abovementioned catalog in Rancher 

Results:
- the catalog is added and synced successfully, invalid templates are ignored
- Rancher logs show the following message:
```
2020/04/06 23:34:50 [INFO] Updating global catalog test
2020/04/06 23:34:56 [INFO] Catalog sync done. 15 templates created, 0 templates updated, 0 templates deleted, 0 templates failed
```

![image](https://user-images.githubusercontent.com/6218999/78615213-ba1ff100-7825-11ea-9844-8d4fd78fe5ac.png)
